### ISSUE TITLE
Bootswatch themes repo and maintenance

### ISSUE THREAD

**Issue by memeplex** [mkdocs/mkdocs#1056]
I've updated my version of bootswatch mkdocs themes to the latest mkdocs theme and bootswatch (upstream) themes.

Now, I know there is the mkdocs/mkdocs-bootswatch repository but it's outdated both regarding mkdocs and regarding bootswatch. I've [automated (as much as possible)](https://github.com/mkdocs/mkdocs-bootswatch/issues/21) the process of building the entire familiy of themes from the upstream bootswatch themes and w.r.t. the latest mkdocs theme. Thanks to this automation, I keep my themes very up-to-date and they fix most of the issues reported to mkdocs/mkdocs-bootswatch tracker.

A time ago I offered to become the maintainer for mkdocs/mkdocs-bootswatch but get no answer. Despite the automation, keeping the themes up-to-date is no free lunch, so I need to take a decision between:
1. I become the maintainer of mkdocs/mkdocs-bootswatch.
2. I add an entry to the wiki and keep my own version of mkdocs-bootswatch, which will be a sad state of affairs both because of the duplication and because I see no drawback in my version.
3. The current maintainer (I think there isn't one just right now) accepts my version. Note: this is not a simple PR because of some git tagging & branching the automation depends on.
4. I egoistically focus on my own specific need (yeti theme) and forget about this.

What do you prefer?

**Comment by waylan** [comment]
I did a lot of refactoring of template variables in #921 and #1026. Those changes need to be made to the [mkdocs-bootswatch][1], [mkdocs-bootstrap][2], [mkdocs-basic-theme[3], and [theme-template[4] repos. I intending to do PRs for each, but I ran out of time at first, then I forgot about them (as I don't use them). In fact, now that I think about it, that should happen before the next release (adds this to the 0.16 milestone...).

I would expect that PRs are welcome.

[1]: /mkdocs/mkdocs-bootswatch
[2]: /mkdocs/mkdocs-bootstrap
[3]: /mkdocs/mkdocs-basic-theme
[4]: /mkdocs/theme-template

**Comment by memeplex** [comment]
@waylan unless I'm missing something else here all your refactorings should be included in my version of the themes, since they're automatically generated from mkdocs master and bootswatch 3. In fact, using your stuff was the eason I first wrote the gen.sh script. There is a tweaks branch in my repo with the changes going from mkdocs theme to each bootswatch variant, which I carefully updated to include both previous divergences and your improvements. This mainly applies to base.css, the rest of the theme and templates are more or less the same for mkdocs than for bootswatch, except for minor tweaks. Each time I want to update the suite I run gen.sh that clones mkdocs and downloads the bootswatch stuff, then reapplies the tweaks branch on top. If there are any conflicts I manually update and rebase the tweaks branch. All in all it's very streamlined. 

Please take a look at https://github.com/mkdocs/mkdocs-bootswatch/issues/21 and at my fork and tell me if there is something you would like to add or change.

**Comment by memeplex** [comment]
Just to be clear: it's not a PR by PR approach. I asked d0ugal at the time and he agreed an automated approach was desirable. My experience since then has been very positive, since I was able to routinely pick your improvements without much effort. The first time was a PITA though since I had to create the diff, that was a lot of manual work.

**Comment by waylan** [comment]
Interesting. I assumed "automation" meant having one master theme (defined with LESS/SASS/etc) that would then generate all of the various subthemes. Then when any future changes came in, one would simply update the master theme and then run a command to regenerate all the subthemes. In contrast, it seems that you took "automation" to mean update the subthemes from the MkDocs theme. I can see value in both, but I would suspect your script would be much similar if it only needed to update the one master template, then let LESS/SASS/etc do the rest.

Similarly, I expected something similar for the bootstrap themes (one master theme which generated all the subthemes). In fact, Bootstrap even has a defined config file format (JSON) which you can [provide to their site][1] which then generates the custom set of components you need. When Bootstrap gets updated, then you can get your project updated without wading through all the files and long diffs yourself (a few config file tweaks should be all that's needed). Unfortunately, it appears that no one saved the config that was used when the bootstrap themes were first created.

In both cases, I would like to see the master theme in one subdir of the repo and then the generated themes in another subdir of the repo. Of course, the Python package would consist of the generated subdir only. But any changes to the repo would require you to edit the master theme, then run the script to update the subthemes before committing.

[1]: http://getbootstrap.com/customize/

**Comment by memeplex** [comment]
I've considered two options at the time: having a jinja template for base.css and keeping a diff at the git level. I chose the git approach because it's more flexible (any arbitrary change can be done to the base theme) and straightforward (I don't have to design some templating logic upfront). Most changes are related to colors and fonts and I don't see a clear advantage in having a dictionary of colors vs a 10-line diff mostly like -this_color +that_color.

The diffs are very simple, they are long just because of the replaced bootstrap css, but this is trivially automated stuff. The only time you need to unnecessarily (vs. the template approach) go fix theme by theme is when some change in the base theme produces some conflict at git level when cherry-picking the update..tweaks commits. To be sincere, I'd rather edit a couple of conflict-marked lines in a handful of files from time to time than keep templates which has to be manually updated to reflect each of your changes to the base theme (except you want to further generalize the base theme to subsume the bootswatch ones since it's essentially a bootswatch/bootstrap theme). 

If you dislike the approach I still might add an entry to the wiki and rename my themes so that they don't clash the mkdocs-bootswatch package.

**Comment by memeplex** [comment]
One final word: I talk about the diff but it's a git branch tracking the minimal changes from base to target theme. So it's extensible by adding further commits, rebasable to changes in the base theme and documentable in commit comments. I tried to leverage git strengths, it's not just a big once-and-hopefully-for-all diff.

**Comment by waylan** [comment]
I've updated the bootswatch and bootstrap themes for MkDocs v0.16 at mkdocs/mkdocs-bootswatch#23 and mkdocs/mkdocs-bootstrap#8 respectively. Note that I also created a base for the bootswatch themes and automated generation of the individual themes in mkdocs/mkdocs-bootswatch#22. I certainly appreciated being able to only make the changes once for all of the themes.

**Comment by memeplex** [comment]
Thanks waylan, I'm happy to see you have automated the theme generation (and happier than I don't have to maintain my version anymore :) ).

**Comment by memeplex** [comment]
@waylan I have been taking a look at your changes and I have a couple of questions. My fork automatically updates the themes from upstreams (both mkdocs and bootswatch). I see your method just generates the themes from _base but neither updates _base nor the bootswatch sources (besides, you're using an old bootswatch version, there are new themes now and amelia is deprecated). How do you plan to keep _base up to date? Will you manually copy every change you make to mkdocs templates?

**Comment by mstenta** [comment]
@memeplex and @waylan related: https://github.com/mkdocs/mkdocs-bootswatch/issues/7

**Comment by memeplex** [comment]
@mstenta I'm aware of that issue, that's one of the reasons I forked the repo and automated the generation of all themes directly from the sources.

**Comment by waylan** [comment]
I just did the minimum required to get the themes working with the upcoming changes to MkDocs (renamed variables.etc). Any additional changes are optional. I'm not sure what @d0ugal had in mind for the Bootswatch themes. Did he intend them to be updated with every Bootswatch update? What about any changes/additions to the MkDocs theme?

As the themes are now, we can do a release contemporaneously with a MkDocs release (which is almost ready to go). After that, then we can explore allowing the themes to move forward on their own (if that's the intention).

**Comment by d0ugal** [comment]
> I just did the minimum required to get the themes working with the upcoming changes to MkDocs (renamed variables.etc). Any additional changes are optional. I'm not sure what @d0ugal had in mind for the Bootswatch themes. Did he intend them to be updated with every Bootswatch update? What about any changes/additions to the MkDocs theme?

I didn't have a particular plan in mind. These themes existed before I got involved, I just done the minor updates and moved them to this external repo.

I would be happy to see these themes develop and grow and be updated as needed etc. However, it all depends on somebody having the time and motivation for this to happen.

**Comment by memeplex** [comment]
Just to let you know I have updated [my automated bootswatch themes](https://github.com/memeplex/mkdocs-bootswatch) to latest bootswatch 3 and latest mkdocs theme. I again offer you to become the maintainer of the bootswatch themes (using my generator) or, alternatively, to rename them to mkdocs-bootswatch3 or something like that and add an entry to themes list.

**Comment by d0ugal** [comment]
I think we should really consider this. I don't use the Bootswatch themes so I have very little time for them. @waylan what do you think?

**Comment by memeplex** [comment]
Another possibility is to keep waylan's approach but to check if the base file was made a template or not and in the first case replace the theme variables (mostly colors) before copying. In my experience the real PITA is to keep updated each base.css. Then I could keep my separate repo and periodically  download and tweak everything, just as I'm doing now, except that I would be tweaking just the _base directory; I might submit periodical PR ps with the updates to _base to your repo. This combines both approaches strengths and doesn't require me to become an official maintainer. The TODO is: templatize base css. I may do that.

**Comment by memeplex** [comment]
To make the above more concrete:

1. I submit a PR that patches setup.py to render a jinja template in case the source in _base was templatized. This is mainly for base.css, so setup.py could also special case this file instead of implementing a generic hook.

2. I update gen.sh to do the same it's doing now (downloading and updating bootstrap 3 and vanilla mkdocs, patching the themes) but only patching the _base directory. It would be convenient for me if you let me keep gen.sh at mkdocs-bootswatch master branch but, otherwise, I could keep a copy of the script elsewhere.

3. I periodically run my script and submit PRs to mkdocs-bootswatch with the update.

This way:

1. Updating from upstream (bootswatch, mkdocs0) is complete automatic.

2. There is only one base.css to patch (the templatized one in _base).

3. I still keep a branch of theme tweaks to be applied on top of the update each time, which I find very convenient.

4. You don't need to depend on me as an official maintainer but just review and hopefully accept my contributions as PRs.

**Comment by waylan** [comment]
I'm not using those themes either and haven't even bothered to subscribe ("watch" in GitHub parlance) to updates for them. Therefore, I don't get notified of new issues or PRs. I made the updates to them to match the changes I made to MkDocs, but that is the extent of anything I have done with them. It may make sense to hand maintenance over to someone who actually uses them.

Btw, we probably should have done a release of all of the external theme when we released 0.16 of MkDocs. We had numerous backward incomparable changes that were theme related. We are still in the smooth transition stage so it doesn't actually break the themes yet (just raises a bunch of warnings). But we should be setting the example for third party theme maintainers. The themes were updated and should be ready to release (at least to match the significant 0.16 changes). However both of us completely forgot about it when we did the release... which, to me, seems like a good reason to pass of maintenance to someone else.
